#1294 update docs for azure (#45)
edgelesssys/docs.edgeless.systems: edgelessdb/getting-started/quickstart-azure.md

# Quickstart: Microsoft Azure

> In this guide, you will set up EdgelessDB using the Microsoft Azure Marketplace.

Reviewer: This wording is taken from the other quickstarts, which in turn were taken from the marblerun quickstarts.

## Get the machine
Visit the Azure Marketplace and get [EdglessDB](https://azuremarketplace.microsoft.com/en-us/marketplace/apps/edgelesssystems.edb?tab=Overview).

> ![azure_create](../_assets/azure/azure_get.png)

Reviewer: Seems like docsify ignores relativePath config for images, so path must be `(_media/azure_get.png)`. Unfortunately, the preview in vscode then no longer works. So sad.

> ![azure_create](../_assets/azure/azure_get.png)

Reviewer: * Pls use reasonable alt texts for the images
* Rename azure_wizard_edited.png to azure_wizard.png
* azure_create_edited.png is unused

> ![azure_create](../_assets/azure/azure_get.png)

Reviewer 2: > Seems like docsify ignores relativePath config for images

We are using a [custom render function for images](https://github.com/edgelesssys/docs.edgeless.systems/blob/b27e20a29a767cc9789e3bf820b12e5f431e59fa/edgelessdb/index.html#L58), but I guess it should work for `relativePath`, since the `relativePath` option was deleted from the MarbleRun `index.html` at a later point. 
I can't remember why we deleted this option for MarbleRun, but we should try to restore consistency between those index files to prevent such mistakes.

After that you have to configure the virtual machine in the creation wizard.

> !> Important: To experience the full potential of EDB we recommend to choose a virtual machine size of the [DCsv2-series](https://docs.microsoft.com/en-us/azure/virtual-machines/dcv2-series). This guarantees, that your virtual machine supports Intel SGX.

Reviewer: This sounds like EDB (btw pls always write EdgelessDB in public texts) would work on other machines. But I guess it won't start, will it?

> The resource group you deploy your machine to has to be created in one of the regions where the DCsV2 sizes are [available](https://azure.microsoft.com/de-de/global-infrastructure/services/?products=virtual-machines&regions=all).

Reviewer 2: ```suggestion
Created the resource group for your machine in [one of the regions](https://azure.microsoft.com/de-de/global-infrastructure/services/?products=virtual-machines&regions=all) where the DCsv2 series is available.
```

The value for the `Virtual machine name` can be freely selected.
The value for the `Username` **must** be set to _edb_.

> Finally press the blue button `Review + Create` and after that the `Create` button to finish the setup process.

Reviewer 2: ```suggestion
Finally, press the blue `Review + Create` button and then the `Create` button to finish the setup process.
```

![azure_wizard](../_assets/azure/azure_wizard_edited.png)

## Get the IP

> After you have created the machine you need the public IP address to interact with it.
> For that naviagte to the resource group your virtual machine was created in and select the virtual machine.
> The public IP address is displayed on the right hand side in the `Essentials` section.

Reviewer 2: ```suggestion
After you have created the machine, get the public IP address of the database you just created, so you can interact with it.
Navigate to the resource group your virtual machine was created in and select the virtual machine.
The public IP address is displayed on the right hand side in the `Essentials` section.
```

> Alternatively you can use the [azure-cli](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli) to get the public IP address:

Reviewer 2: ```suggestion
Alternatively, you can use the [azure-cli](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli) to get the public IP address:
```

```bash
 az vm show -d -g <resource-group> -n <vm-name> --query publicIps -o tsv
```

## Generate certificates and create a manifest
You will now create a manifest that defines a root user. This user is authenticated by an X.509 certificate.

Generate a certificate authority (CA) and a corresponding user certificate:
```bash
openssl req -x509 -newkey rsa -nodes -subj '/CN=My CA' -keyout ca-key.pem -out ca-cert.pem
openssl req -newkey rsa -nodes -subj '/CN=rootuser' -keyout key.pem -out csr.pem
openssl x509 -req -CA ca-cert.pem -CAkey ca-key.pem -CAcreateserial -in csr.pem -out cert.pem
```

Escape the line breaks of the CA certificate:
```bash
awk 1 ORS='\\n' ca-cert.pem
```

> Create a file `manifest.json` with the following contents:

Reviewer 2: ```suggestion
Create a `manifest.json` file with the following content:
```

Reviewer: AFAIK contents is correct here.
Regarding "file manifest.json" vs "manifest.json file": Not sure, but to me the former is short for "a file named manifest.json" and the latter would emphasize the kind of file (e.g., system file, JPEG file), but we would probably want to emphasize the name of the file.

```json
{
    "sql": [
        "CREATE USER root REQUIRE ISSUER '/CN=My CA' SUBJECT '/CN=rootuser'",
        "GRANT ALL ON *.* TO root WITH GRANT OPTION"
    ],
    "ca": "-----BEGIN CERTIFICATE-----\n...\n-----END CERTIFICATE-----\n"
}
```

`sql` is a list of SQL statements that define the initial state of the database. The two statements above create a root user that is authenticated by the user certificate you just generated.

Replace the value of `ca` with the escaped content of `ca-cert.pem`.

## Verify your EdgelessDB instance

> Before you can trust your EdgelessDB instance, you first need to verify that it is in a good shape. You can use the [Edgeless Remote Attestation (era)](https://github.com/edgelesssys/era) tool for this. If you're just getting started, you may also skip this part.

Reviewer 2: ```suggestion
Before you can trust your EdgelessDB instance, you first need to verify its confidentiality. You can use the [Edgeless Remote Attestation (era)](https://github.com/edgelesssys/era) tool for this. If you're just getting started, you may also skip this part.
```

Reviewer 2: > If you're just getting started, you may also skip this part.

It think the validation with era is quite simple and rather essential for EdgelessDB, why would you want to skip it? If you want to keep this line, please add a warning that this is unsafe and EdgelessDB should always be attested in production.

Reviewer: > 
> 
> > If you're just getting started, you may also skip this part.
> 
> It think the validation with era is quite simple and rather essential for EdgelessDB, why would you want to skip it? If you want to keep this line, please add a warning that this is unsafe and EdgelessDB should always be attested in production.

Felix introduced this sentence.

Reviewer 2: > Felix introduced this sentence.

@thomasten For what reason?

Reviewer: I'd guess he disagrees that "the validation with era is quite simple" for first-time users ;)

Once you've installed `era`, you can get the attested root certificate of your EdgelessDB instance as follows:
```bash
wget https://github.com/edgelesssys/edgelessdb/releases/latest/download/edgelessdb-sgx.json
era -c edgelessdb-sgx.json -h <your-azure-ip>:8080 -output-root edb.pem -skip-quote
```

> Here, `edgelessdb-sgx.json` contains the expected properties of your EdgelessDB instance. However, in simulation mode, you need to skip the actual verification of the properties via the `-skip-quote` option.

Reviewer 2: Is this quickstart using any kind of simulation mode? This wasn't mentioned previously! Why would we use simulation mode if we just created a confidential database on an SGX enabled machine?

Reviewer: @leongross pls use the SGX quickstart as template for this page, not the simulation quickstart.
